### PR TITLE
replace Lock with RLock for lruCache

### PR DESCRIPTION
**Pull request by adairxie**
**Please provide a description of this PR:**
Hello!
    I am optimizing the code for my company， then i found that the get method of XdsCache is using Lock not the RLock，which makes me wondoer。I change it and test it, this the result:
    For Lock:
        image
   For RLock:
       image

   It saves about 10ns/op on my computer.

### CODE REVIEW

File: pilot/pkg/model/typed_xds_cache_test.go [hunk 1]
```diff
@@ -440,3 +440,31 @@ func testEvictQueueMemoryLeak(t *testing.T, f string) {
 		}
 	}
 }
+
+func BenchmarkCacheGet(t *testing.B) {
```
Comment (wulianglongrd) [inline pilot/pkg/model/typed_xds_cache_test.go:444]: I think this benchmark is unnecessary, the performance improvement seems undoubted, and besides, we already have relevant test, see `BenchmarkCache`.

**Comment by wulianglongrd** [comment]
The internal LRU implementation has data race.

```
{Failed  ==================
WARNING: DATA RACE
Read at 0x00c002a35e30 by goroutine 18725:
  github.com/hashicorp/golang-lru/v2/internal.(*LruList[go.shape.uint64,go.shape.struct { istio.io/istio/pilot/pkg/model.value *github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3.Resource; istio.io/istio/pilot/pkg/model.token istio.io/istio/pilot/pkg/model.CacheToken; istio.io/istio/pilot/pkg/model.dependentConfigs []istio.io/istio/pilot/pkg/model.ConfigHash }]).MoveToFront()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/internal/list.go:137 +0x164
  github.com/hashicorp/golang-lru/v2/simplelru.(*LRU[go.shape.uint64,go.shape.struct { istio.io/istio/pilot/pkg/model.value *github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3.Resource; istio.io/istio/pilot/pkg/model.token istio.io/istio/pilot/pkg/model.CacheToken; istio.io/istio/pilot/pkg/model.dependentConfigs []istio.io/istio/pilot/pkg/model.ConfigHash }]).Get()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/simplelru/lru.go:73 +0xe6
  github.com/hashicorp/golang-lru/v2/simplelru.(*LRU[uint64,istio.io/istio/pilot/pkg/model.cacheValue]).Get()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/simplelru/lru.go:71 +0x73
  istio.io/istio/pilot/pkg/model.(*lruCache[go.shape.uint64]).get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/typed_xds_cache.go:289 +0x114
  istio.io/istio/pilot/pkg/model.(*lruCache[go.shape.uint64]).Get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/typed_xds_cache.go:282 +0x4e
  istio.io/istio/pilot/pkg/model.(*lruCache[uint64]).Get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/typed_xds_cache.go:281 +0x17
  istio.io/istio/pilot/pkg/model.XdsCacheImpl.Get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/xds_cache.go:150 +0x112
  istio.io/istio/pilot/pkg/model.(*XdsCacheImpl).Get()
      <autogenerated>:1 +0xe5
  istio.io/istio/pilot/pkg/networking/core.(*ClusterBuilder).getAllCachedSubsetClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster_builder.go:645 +0x210
  istio.io/istio/pilot/pkg/networking/core.(*ConfigGeneratorImpl).buildOutboundClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster.go:308 +0x56b
  istio.io/istio/pilot/pkg/networking/core.(*ConfigGeneratorImpl).buildClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster.go:218 +0x1b5
  istio.io/istio/pilot/pkg/networking/core.(*ConfigGeneratorImpl).BuildClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster.go:64 +0x276
  istio.io/istio/pilot/pkg/xds.CdsGenerator.Generate()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/cds.go:127 +0x6f
  istio.io/istio/pilot/pkg/xds.(*CdsGenerator).Generate()
      <autogenerated>:1 +0x84
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushXds()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/xdsgen.go:120 +0x947
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushConnection()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:490 +0x216
  istio.io/istio/pilot/pkg/xds.(*Connection).Push()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:130 +0x67
  istio.io/istio/pkg/xds.Stream()
      /home/prow/go/src/istio.io/istio/pkg/xds/server.go:285 +0x3b5
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).Stream()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:238 +0xc64
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).StreamAggregatedResources()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:183 +0x3a
  github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3._AggregatedDiscoveryService_StreamAggregatedResources_Handler()
      /home/prow/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.13.1-0.20240823165802-4363a624d376/envoy/service/discovery/v3/ads_grpc.pb.go:137 +0xe5
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1673 +0x1fd9
  google.golang.org/grpc.(*Server).handleStream()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1794 +0x17c4
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1029 +0x158

Previous write at 0x00c002a35e30 by goroutine 18744:
  github.com/hashicorp/golang-lru/v2/internal.(*LruList[go.shape.uint64,go.shape.struct { istio.io/istio/pilot/pkg/model.value *github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3.Resource; istio.io/istio/pilot/pkg/model.token istio.io/istio/pilot/pkg/model.CacheToken; istio.io/istio/pilot/pkg/model.dependentConfigs []istio.io/istio/pilot/pkg/model.ConfigHash }]).move()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/internal/list.go:117 +0x366
  github.com/hashicorp/golang-lru/v2/internal.(*LruList[go.shape.uint64,go.shape.struct { istio.io/istio/pilot/pkg/model.value *github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3.Resource; istio.io/istio/pilot/pkg/model.token istio.io/istio/pilot/pkg/model.CacheToken; istio.io/istio/pilot/pkg/model.dependentConfigs []istio.io/istio/pilot/pkg/model.ConfigHash }]).MoveToFront()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/internal/list.go:141 +0x175
  github.com/hashicorp/golang-lru/v2/simplelru.(*LRU[go.shape.uint64,go.shape.struct { istio.io/istio/pilot/pkg/model.value *github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3.Resource; istio.io/istio/pilot/pkg/model.token istio.io/istio/pilot/pkg/model.CacheToken; istio.io/istio/pilot/pkg/model.dependentConfigs []istio.io/istio/pilot/pkg/model.ConfigHash }]).Get()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/simplelru/lru.go:73 +0xe6
  github.com/hashicorp/golang-lru/v2/simplelru.(*LRU[uint64,istio.io/istio/pilot/pkg/model.cacheValue]).Get()
      /home/prow/go/pkg/mod/github.com/hashicorp/golang-lru/v2@v2.0.7/simplelru/lru.go:71 +0x73
  istio.io/istio/pilot/pkg/model.(*lruCache[go.shape.uint64]).get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/typed_xds_cache.go:289 +0x114
  istio.io/istio/pilot/pkg/model.(*lruCache[go.shape.uint64]).Get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/typed_xds_cache.go:282 +0x4e
  istio.io/istio/pilot/pkg/model.(*lruCache[uint64]).Get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/typed_xds_cache.go:281 +0x17
  istio.io/istio/pilot/pkg/model.XdsCacheImpl.Get()
      /home/prow/go/src/istio.io/istio/pilot/pkg/model/xds_cache.go:150 +0x112
  istio.io/istio/pilot/pkg/model.(*XdsCacheImpl).Get()
      <autogenerated>:1 +0xe5
  istio.io/istio/pilot/pkg/networking/core.(*ClusterBuilder).getAllCachedSubsetClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster_builder.go:645 +0x210
  istio.io/istio/pilot/pkg/networking/core.(*ConfigGeneratorImpl).buildOutboundClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster.go:308 +0x56b
  istio.io/istio/pilot/pkg/networking/core.(*ConfigGeneratorImpl).buildClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster.go:218 +0x1b5
  istio.io/istio/pilot/pkg/networking/core.(*ConfigGeneratorImpl).BuildClusters()
      /home/prow/go/src/istio.io/istio/pilot/pkg/networking/core/cluster.go:64 +0x276
  istio.io/istio/pilot/pkg/xds.CdsGenerator.Generate()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/cds.go:127 +0x6f
  istio.io/istio/pilot/pkg/xds.(*CdsGenerator).Generate()
      <autogenerated>:1 +0x84
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushXds()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/xdsgen.go:120 +0x947
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushConnection()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:490 +0x216
  istio.io/istio/pilot/pkg/xds.(*Connection).Push()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:130 +0x67
  istio.io/istio/pkg/xds.Stream()
      /home/prow/go/src/istio.io/istio/pkg/xds/server.go:285 +0x3b5
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).Stream()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:238 +0xc64
  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).StreamAggregatedResources()
      /home/prow/go/src/istio.io/istio/pilot/pkg/xds/ads.go:183 +0x3a
  github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3._AggregatedDiscoveryService_StreamAggregatedResources_Handler()
      /home/prow/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.13.1-0.20240823165802-4363a624d376/envoy/service/discovery/v3/ads_grpc.pb.go:137 +0xe5
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1673 +0x1fd9
  google.golang.org/grpc.(*Server).handleStream()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1794 +0x17c4
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1029 +0x158

Goroutine 18725 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1040 +0x224
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/internal/transport/http2_server.go:630 +0x3941
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/internal/transport/http2_server.go:671 +0x406
  google.golang.org/grpc.(*Server).serveStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1023 +0x69b
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:959 +0x86

Goroutine 18744 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1040 +0x224
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/internal/transport/http2_server.go:630 +0x3941
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/internal/transport/http2_server.go:671 +0x406
  google.golang.org/grpc.(*Server).serveStreams()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:1023 +0x69b
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /home/prow/go/pkg/mod/google.golang.org/grpc@v1.65.0/server.go:959 +0x86
==================
```

**Comment by hzxuzhonghu** [comment]
The panic is expected, hashicorp LRU cache does not gurantee concurrent safe, `Get ` is not just a read op, it has to maitain the element that is being read

**Comment by ramaraochavali** [comment]
I remember we discussed this and consciously left it with Lock intead of RLock

File: pilot/pkg/model/typed_xds_cache_test.go (hunk 1 above)
Comment (adairxie) [reply pilot/pkg/model/typed_xds_cache_test.go]: Ok, I delete it.

**Comment by adairxie** [comment]
Thank you all, guys. I have a question that why not use a normal map instead of lrucache?

**Comment by zirain** [comment]
> I remember we discussed this and consciously left it with Lock intead of RLock

@adairxie can you update the comment from @ramaraochavali , so that no one will try to do same thing in the future.

**Comment by adairxie** [comment]
> I remember we discussed this and consciously left it with Lock intead of RLock
@ramaraochavali You are right, RLock will cause a race. I will close this pr, thank you!

**Pull request closed without merging**